### PR TITLE
Reorder tests in maybe_downcast_numeric

### PR DESCRIPTION
**Pull request by MichaelTiemannOSC**
The comment `# if we have any nulls, then we are done` is not consistent with the test `if isna(arr).any()` because `arr` is constructed only from the first element (`r[0]`) not the full ravel'd list of values.  Moreover, calling `np.array()` on some random type can have surprising consequences.

So instead, do the early-out test as intended, just using `r[0]` without going through `np.array()`.  Then test other things about `r[0]`.  Only then should we test all the values (and if we have any nulls, then we are done).

See #55824

- [55824] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by MichaelTiemannOSC** [comment]
While there is a conversation going on in #55824, these changes introduce no new functionality nor user-visible changes (other than tolerating Pint's behavior).  All these changes do is remove a dependency on the behavior of `np.array()` that can be accomplished as well by other means, and to make the implementation match the comments regarding the results of encountering NaN/NA values in the values to be possibly downcasted.

**Comment by WillAyd** [comment]
Cool thanks for the PR. Can you add a test and a whatsnew note?

**Comment by MichaelTiemannOSC** [comment]
Would be happy to, but...the test case would require loading Pint.  Last time I tried to do some test cases with alien packages I got back a big "No thanks!" from the pandas maintainers.  Also, another PR I submitted (enabling many complex test cases) rejected the WhatsNew changes I wrote because there were no user-visible changes.  Would you like to guide me further?

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -357,18 +357,20 @@ def trans(x):
...
             return trans(result).astype(dtype)
 
-        # do a test on the first element, if it fails then we are done
         r = result.ravel()
-        arr = np.array([r[0]])
 
-        if isna(arr).any():
-            # if we have any nulls, then we are done
+        if isna(r[0]):
+            # do a test on the first element, if it fails then we are done
```
Comment (rhshadrach) [inline pandas/core/dtypes/cast.py:363]: This test on the first element goes back to https://github.com/pandas-dev/pandas/commit/dc73315d557874a7f162dcf0bc3ee106f9338ec5#diff-fb8a9c322624b0777f3ff7e3ef8320d746b15a2a0b80b7cab3dfbe2e12e06daa in `core.common`. It appears to me it is no longer necessary.

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -357,18 +357,20 @@ def trans(x):
...
             return result
 
         elif not isinstance(r[0], (np.integer, np.floating, int, float, bool)):
             # a comparable, e.g. a Decimal may slip in here
             return result
 
+        if isna(r).any():
+            # if we have any nulls, then we are done
+            return result
```
Comment (rhshadrach) [inline pandas/core/dtypes/cast.py:372]: Isn't this duplicative of `notna(result).all()` on L376 below?

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (MichaelTiemannOSC) [reply pandas/core/dtypes/cast.py:363]: So I constructed a simple test case and I think I understand what the earlier code was trying to do, which was to `ravel` only the first element of `result`.  Consider:

```
import numpy as	np
import pandas as pd

aa = np.array([1.0, np.nan, 2.0])
bb = np.array([aa, aa])

print(aa, pd.isna(aa))
# [ 1. nan  2.] [False  True False]
print(bb, pd.isna(bb))
#  [[ 1. nan  2.] [ 1. nan  2.]] [[False  True False] [False  True False]]
print(pd.isna(bb).any())
# True
```
The fact that `pd.isna(bb).any()` sees the nulls inside the sub-arrays means we don't need to ravel the whole thing to look for nulls.  The reason to ravel, instead, is to look at the type of the elements, and we only need to ravel the first element to look into that.  (We have already tested that there are elements to ravel.)

File: pandas/core/dtypes/cast.py (hunk 2 above)
Comment (MichaelTiemannOSC) [reply pandas/core/dtypes/cast.py:372]: Agree.  Will patch and re-submit.

Comment (MichaelTiemannOSC) [reply pandas/core/dtypes/cast.py]: Done...

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -357,15 +357,13 @@ def trans(x):
...
-        r = result.ravel()
-        arr = np.array([r[0]])
-
-        if isna(arr).any():
-            # if we have any nulls, then we are done
-            return result
-
-        elif not isinstance(r[0], (np.integer, np.floating, int, float, bool)):
+        if isinstance(result, np.ndarray):
+            element = result[0]
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:361]: would `result.item(0)` work here? might avoid potential copies from ravel

**Comment by MichaelTiemannOSC** [comment]
So to wrap this up...I don't think it's easy to add a test case because that would involve installing Pint, which might not be very CI/CD-friendly.  If there's a good way to do that, please point me to a pattern I can follow.  It's easy for me to describe the change in CHANGES, but its not really user-visible.  But because there is an issue (though no test case), I could write it up anyway.  Thoughts?

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -357,15 +357,11 @@ def trans(x):
...
-        # do a test on the first element, if it fails then we are done
-        r = result.ravel()
-        arr = np.array([r[0]])
-
-        if isna(arr).any():
-            # if we have any nulls, then we are done
-            return result
-
-        elif not isinstance(r[0], (np.integer, np.floating, int, float, bool)):
+        if isinstance(result, np.ndarray):
```
Comment (rhshadrach) [inline pandas/core/dtypes/cast.py:360]: In what case are we dealing with something that isn't a NumPy array?

Comment (MichaelTiemannOSC) [reply pandas/core/dtypes/cast.py:360]: `pd.interval_range` seems to be the common organic reason.  `tests/dtypes/cast/test_downcast` file triggers the case by explicitly creating `ndarray` objects.

Comment (rhshadrach) [reply pandas/core/dtypes/cast.py:360]: It looks like we get here with Series, which isn't part of the type-annotations. That should probably be fixed, but outside the scope here.

**Comment by rhshadrach** [comment]
@WillAyd 

> Can you add a test and a whatsnew note?

With the current implementation in this PR, I wouldn't expect any behavior changes. You good here?

It does change for Pint because there `np.array([e])` will raise for a Pint object, but I don't think we should be testing for this.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @MichaelTiemannOSC

File: pandas/core/dtypes/cast.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: My suggestion here turns out to cause a different problem bc .item(0) is wrong for dt64 ndarrays

```
result = np.arange(5).view("M8[ns]")
>>> result.item(0)
0
```

**Pull request merged**